keps/sig-auth/apiserver_auth_to_webhooks.md

```diff
@@ -0,0 +1,128 @@
+
+## Motivation
+
+Outgoing webhooks from the apiserver such as the ones found in [dynamic admission control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) and [dynamic audit](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#dynamic-backend) 
+suffer from a lack of easily configurable authentication. Currently, Dynamic Admission webhooks provide a mechanism for plugin authentication by a kubeconfig provisioned on the host. The intention of this KEP is to provide a simpler means for the receiving 
+server to authenticate the apiserver and its aggregates using the [Authentication API](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go).
+
+### Goals
+* A simple means of authenticating apiserver clients.
+
+### Non-Goals
+* Providing all the authentication schemes found in the current kubeconfig.
+
+## Proposal
+
+We propose a simple mechanism for authenticating webhooks using the token [Authentication API](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go). The shared 
+[webhook client](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go) will be parameterized to optionally enhance every outgoing request with a token obtained from a  
+[Token Request](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L112). The receiving server can then check that token using a [Token Review](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L45).
+
+### User Stories
+
+#### Story 1
+I am a cluster administrator using the dynamic auditing feature and want to make sure the logs I receive are from the apiserver.
+
+#### Story 2
+I am a plugin developer and want to easily authenticate the apiserver.
+
+### Implementation Details/Notes/Constraints
+
+A new struct will be added to the [client config](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/webhook/client.go#L40) for outgoing webhooks:
+
+```go
+type AuthInfo struct {
+  ProvisionToken  bool
+}
+
+type ClientConfig struct {
+  Name     string
+  URL      string
+  CABundle []byte
+  AuthInfo *AuthInfo
+  Service  *ClientConfigService
+}
+```
+If enabled the client will provision a token and enhance the outgoing request with an auth header:
+```
+Authorization: bearer <token>
+```
+
+The client will check and refresh the token when necessary. The server can now check the token using a 
+[TokenReview](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/authentication/types.go#L45).
+
+The `AuthInfo` struct will also be added to the [ClientConfig](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/auditregistration/v1alpha1/types.go#L134) in the auditregistration API.
+
+The token audience would be that of the webhook name. The receiving server could verify that it is the intended 
```

Comment:
Is this audience embedded in the token? If so, does the receiving Webhook extract the audience from the token, or is it relying on a review to yield a `TokenReviewStatus` with the "Audiences" field populated?

It wasn't clear on first reading, but if I understand this idea correctly, the Webhook server doesn't trust its clients, but it trusts some API server for token reviews—even though that same API server might in fact be the client of these Webhooks.

Thinking about those paired connections, it seems like client certificate authentication would be easier to understand: Since the Webhook server already knows whether or not to trust the API server by way of its server certificate, the Webhook server could verify that clients that connect present a certificate signed by the same trusted CA that covers the API server's certificate. Granted, it's less practical to limit the validity period and audience scope in such X.509 certificates, but it is worth calling out why we need two network connections here to authenticate a client expected to be one and the same as a server we trust.

Reply:
> Is this audience embedded in the token? If so, does the receiving Webhook extract the audience from the token, or is it relying on a review to yield a `TokenReviewStatus` with the "Audiences" field populated?

The webhook submits the audiences it identifies as in the TokenReviewSpec. TokenReviewStatus.auds will return all audiences that match. The webhook shouldn't need to introspect the token.

> it seems like client certificate authentication would be easier to understand

This is possible today via the webhook kubeconfig, but requires reconfiguring the kube-apiserver. I think we want to support delegated token auth in addition. I could see AuthInfo including an option that does Client TLS with a static client cert shared across all webhooks.

Reply:
Does the identity (user name, groups, user ID) yielded by the `TokenReviewStatus` matter to the Webhook, or do you only anticipate it caring about the expected audience intersection?

Reply:
It sounds like you're assuming that the Webhook server will understand the values in the `UserInfo` fields. Where does the `UserInfo` content come from? I mean, I understand it's derived from the token content, but this proposal doesn't say which entity creates the token, or what goes into it.

Reply:
An important use case of this feature would be to have webhooks that are shared across multiple clusters, in particular if the webhook is deployed as a cloud function or on a server with proper DNS certs ( to avoid the horrible CA-cert patching ). 

It would be ideal if the UserInfo would include some info about the cluster making the call, for example in GKE some reference to the project ID/zone/cluster name.